Allow overriding of color order by LED pixel range.

justfalter opened the pull request:
Adds a means for overriding color order for ranges of LED pixels. Default color order is still managed at the GPIO level.  This is meant to fix #2462. Up to 10 color order overrides defined.

- [x] Use `ColorOrderMap` to hold color order information.
- [x] Serialization of config to/from filesystem.
- [x] Modify `handleSettingsSet` to take a configuration
- [x] Modify `xml.cpp` to serialize settings for front-end
- [x] Modify LED settings page to allow for configuration of color order map.

### Performance: 

I figured watching the FPS would be the best way to know if my changes negatively impacted performance. Hopefully the reported FPS is accurate. 

My methodology:
- 1 segment for all 1191 LEDs
- Disabled network discovery.
- Effect: Pride 2015, Speed 255, Effect 255
- Measure FPS 10 times over 10 seconds:
```
#!/bin/bash
num_samples=10
total=0
for i in $(seq 1 ${num_samples}); do
  sleep 1
  fps=$(curl -s "${WLED_BASE_URL}/json/info"|jq '.leds.fps')
  total=$(($total + $fps))
  echo "$i: ${fps}"
done

avg=$(echo "${total}/${num_samples}" | bc -l)
echo "Average: ${avg}"
```

Results:
| Env | FPS | 
|-|-|
| master (736053e24e4933148bd28b6fb2d3720005de9c1e) | 40.4 |
| PR - 0 overrides | 40.4 |  
| PR - 1 overrides | 40.9 |  
| PR - 2 overrides |  40.8 |
| PR - 3 overrides |  39.7 |
| PR - 4 overrides |  39.2 |
| PR - 5 overrides |  40.3 |
| PR - 6 overrides |  40.9 |
| PR - 7 overrides |  40.7 |
| PR - 8 overrides |  39 |
| PR - 9 overrides |  38.9 |
| PR - 10 overrides | 39 |

It appears that performance is comparable between master and 0 overrides. 10 overrides appears to be 3.5% off of the base case.

I ran this several times and, while there was some variability, I'm not really sure why things would vary *up* and *down* like they had. Perhaps it has to do with how FPS is calculated, how the busses are updated, or maybe there are other timing/interrupts occurring.  I thought maybe it was an alignment issue with the `ColorOrderMapEntry` struct, manually padding made no difference. The compiler should take care of aligning things, anyway.  Regardless, performance seems pretty close to what is expected.

### Config

Color order overrides are managed on the LED Settings page:
![image](https://user-images.githubusercontent.com/120545/147860521-9a4cc9c3-aee6-4d9c-932a-5da23aac689e.png)

The above configuration appears within cfg json as:
```
{
  "hw": { "com":[{"start":253,"len":90,"order":2},{"start":696,"len":90,"order":2}] }
}
```

justfalter committed:
Overridable color order

- Use `ColorOrderMap` to hold optional color order overrides for ranges
  of LEDs.
- Serialization of config to/from filesystem is complete.
- Back-end configuration is complete.
- TODO: front-end changes to the LED settings page.

justfalter commented:
The following will add a GPIO with 200 LEDs on it, defaulting to BRG color order, and overriding the color order for pixels 100-200 to be RGB.

```
curl 'http://192.168.55.58/settings/leds?' \
  -H 'Connection: close' \
  -H 'Origin: http://wled-testy.local' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -H 'Referer: http://wled-testy.local/settings/leds?' \
  --data-raw 'MA=850&LAsel=50&LA=0&LT0=22&CO0=2&LC0=200&L00=18&L10=&L20=&L30=&L40=&BT0=0&BE0=2&BT1=-1&BE1=0&BT2=-1&BE2=0&BT3=-1&BE3=0&TT=32&IR=-1&IT=0&data=&RL=12&RM=on&BO=on&CA=128&BP=0&GC=on&BF=100&TF=on&TD=700&TL=60&TB=0&TW=1&AW=3&CB=0&PB=0&data2=&XS0=100&XC0=100&XO0=1' \
  --insecure
```

justfalter committed:
Add Color order override settings

- Adds color order override section to settings page.

justfalter commented:
@Aircoookie any thoughts on this?

blazoncek commented:
It has been considered but is waiting for peer review. Please wait.

blazoncek commented on wled00/bus_manager.h:
```diff
@@ -73,6 +73,62 @@ struct BusConfig {
+    return &(_mappings[n]);
+  }
+
+  inline uint8_t getPixelColorOrder(uint16_t pix, uint8_t defaultColorOrder) const {
```
I am fond of inline functions but I don't see the benefit here.

justfalter replied:
My thinking was that, given that `getPixelColorOrder()`  is called each time a pixel's color is set, inlining `getPixelColorOrder()` would improve performance by eliminating the overhead of having to call and return from the function.

blazoncek commented on wled00/const.h:
```diff
@@ -39,6 +39,8 @@
   #endif
 #endif
 
+#define WLED_MAX_COLOR_ORDER_MAPPINGS 10
```
I would add different default value (5) for ESP8266 to reduce memory pressure and CPU cycles.

justfalter replied:
Done.

blazoncek commented on wled00/xml.cpp:
```diff
@@ -421,6 +421,17 @@ void getSettingsJS(byte subPage, char* dest)
       oappend(SET_F("mA\";"));
     }
 
+    oappend(SET_F("resetCOM();"));
```
Add a value for maximum number of maps.

blazoncek commented on wled00/data/settings_leds.htm:
```diff
@@ -347,6 +347,50 @@
+      btnCOM(i-1);
+    }
+
+    function resetCOM() {
```
Use this function to get maximum number of maps from ESP (see below).

blazoncek commented on wled00/data/settings_leds.htm:
```diff
@@ -347,6 +347,50 @@
+    }
+
+    function btnCOM(i) {
+      gId("com_add").style.display = (i<10) ? "inline":"none";
```
Use variable maximum number of maps.

blazoncek commented on wled00/set.cpp:
```diff
@@ -136,6 +136,18 @@ void handleSettingsSet(AsyncWebServerRequest *request, byte subPage)
     }
 
+    ColorOrderMap com = {};
+    for (uint8_t s = 0; s < WLED_MAX_COLOR_ORDER_MAPPINGS; s++) {
```
What if there are entries missing? It would be better to break instead of adding 0s.

justfalter replied:
Fair point, thank you!

blazoncek reviewed (approved):
Code reviewed (but not yet tested) and gets a go from my side (with minor adjustments suggested).

justfalter committed:
PR Feedback

- Limit max number of color order overrides to 5 on ESP8266
- Only append color overrides if they were provided in the POST of LED
  settings.

blazoncek reviewed (approved):
I hope to have time to pull this into my fork and test this weekend.
Otherwise it looks cool and up to current standards in WLED.

blazoncek commented:
Works wonderfully!
@Aircoookie I recommend we merge this.
